state/machine_linklayerdevices.go

```diff
@@ -1049,276 +1048,19 @@ func (m *Machine) SetDevicesAddressesIdempotently(devicesAddresses []LinkLayerDe
-		out = append(out, fmt.Sprintf("%q", space))
-	}
-	return strings.Join(out, ", ")
-}
-
-// SetContainerLinkLayerDevices sets the link-layer devices of the given
-// containerMachine, setting each device linked to the corresponding
-// BridgeDevice of the host machine. It also records when one of the
-// desired spaces is available on the host machine, but not currently
-// bridged.
-// TODO(jam): 2016-12-28 Move this off of machine
-func (m *Machine) SetContainerLinkLayerDevices(containerMachine *Machine) error {
-	containerSpaces, devicesPerSpace, err := m.findSpacesAndDevicesForContainer(containerMachine)
-	if err != nil {
-		return errors.Trace(err)
-	}
-	logger.Debugf("for container %q, found host devices spaces: %s",
-		containerMachine.Id(), devicesPerSpace)
-
-	spacesFound := set.NewStrings()
-	devicesByName := make(map[string]*LinkLayerDevice)
-	bridgeDeviceNames := make([]string, 0)
-
-	for spaceName, hostDevices := range devicesPerSpace {
-		for _, hostDevice := range hostDevices {
-			deviceType, name := hostDevice.Type(), hostDevice.Name()
-			if deviceType == BridgeDevice {
-				devicesByName[name] = hostDevice
-				bridgeDeviceNames = append(bridgeDeviceNames, name)
-				spacesFound.Add(spaceName)
-			}
-		}
-	}
-	missingSpace := containerSpaces.Difference(spacesFound)
-	if len(missingSpace) > 0 {
-		logger.Warningf("container %q wants spaces %s could not find host %q bridges for %s, found bridges %s",
-			containerMachine.Id(), quoteSpaces(containerSpaces),
-			m.Id(), quoteSpaces(missingSpace), bridgeDeviceNames)
-		return errors.Errorf("unable to find host bridge for space(s) %s for container %q",
-			quoteSpaces(missingSpace), containerMachine.Id())
-	}
-
-	sortedBridgeDeviceNames := network.NaturallySortDeviceNames(bridgeDeviceNames...)
-	logger.Debugf("for container %q using host machine %q bridge devices: %v",
-		containerMachine.Id(), m.Id(), sortedBridgeDeviceNames)
-	containerDevicesArgs := make([]LinkLayerDeviceArgs, len(bridgeDeviceNames))
-
-	for i, hostBridgeName := range sortedBridgeDeviceNames {
-		hostBridge := devicesByName[hostBridgeName]
-		containerDevicesArgs[i] = LinkLayerDeviceArgs{
-			Name:        fmt.Sprintf("eth%d", i),
-			Type:        EthernetDevice,
-			MACAddress:  generateMACAddress(),
-			MTU:         hostBridge.MTU(),
-			IsUp:        true,
-			IsAutoStart: true,
-			ParentName:  hostBridge.globalKey(),
-		}
-	}
-	logger.Debugf("prepared container %q network config: %+v", containerMachine.Id(), containerDevicesArgs)
-
-	if err := containerMachine.SetLinkLayerDevices(containerDevicesArgs...); err != nil {
-		return errors.Trace(err)
-	}
-
-	logger.Debugf("container %q network config set", containerMachine.Id())
-	return nil
+func DefineEthernetDeviceOnBridge(name string, hostBridge *LinkLayerDevice) (LinkLayerDeviceArgs, error) {
+	if hostBridge.Type() != BridgeDevice {
+		return LinkLayerDeviceArgs{}, errors.Errorf("hostBridge must be a Bridge Device not %q", hostBridge.Type())
```

Comment:
getting this would be a programming error. Thoughts on better phrasing?
"can only create an Ethernet device on a bridge not %q"
better?